### PR TITLE
feat: add `napi/argv-complex64`

### LINKED ISSUE #817: [RFC]: Add `@stdlib/napi/argv-complex64`
### Description

This RFC proposes adding C APIs for converting a JavaScript `Complex64` object instance to a native C data type `stdlib_complex64_t`. Should be similar to other `@stdlib/napi/argv-*` APIs.

Package: `@stdlib/napi/argv-complex64`

### Related Issues

None.

### Questions

No.

### Other

No.

### Checklist

- [X] I have read and understood the [Code of Conduct](https://github.com/stdlib-js/stdlib/blob/develop/CODE_OF_CONDUCT.md).
- [X] Searched for existing issues and pull requests.
- [X] The issue name begins with `RFC:`.

### PR DESCRIPTION
**Pull request by Userf1995**
Resolves #817  .

## Description

> What is the purpose of this pull request?

This pull request C APIs for converting a JavaScript Complex64 object instance to a native C data type stdlib_complex64_t

## Related Issues

> Does this pull request have any related issues?

This pull request:

-   resolves #817  

## Questions

> Any questions for reviewers of this pull request?

No.

## Other

> Any other information relevant to this pull request? This may include screenshots, references, and/or implementation notes.

No.

## Checklist

> Please ensure the following tasks are completed before submitting this pull request.

-   [X] Read, understood, and followed the [contributing guidelines][contributing].

* * *

@stdlib-js/reviewers

### CODE REVIEW

File: lib/node_modules/@stdlib/napi/argv-complex64/binding.gyp [hunk 1]
```diff
@@ -0,0 +1,170 @@
+# @license Apache-2.0
+#
+# Copyright (c) 2022 The Stdlib Authors.
```
Comment (Pranavchiku) [inline lib/node_modules/@stdlib/napi/argv-complex64/binding.gyp]:
```suggestion
# Copyright (c) 2024 The Stdlib Authors.
```

Assuming this is copied over.

File: lib/node_modules/@stdlib/napi/argv-complex64/include.gypi [hunk 2]
```diff
@@ -0,0 +1,53 @@
+# @license Apache-2.0
+#
+# Copyright (c) 2022 The Stdlib Authors.
```
Comment (Pranavchiku) [inline lib/node_modules/@stdlib/napi/argv-complex64/include.gypi]:
```suggestion
# Copyright (c) 2024 The Stdlib Authors.
```

File: lib/node_modules/@stdlib/napi/argv-complex64/include/stdlib/napi/argv_complex64.h [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*/
+
+#ifndef STDLIB_NAPI_ARGV_COMPLEX64_H
+#define STDLIB_NAPI_ARGV_COMPLEX64_H
+
+#include "stdlib/napi/argv.h"
+#include "stdlib/assert/napi/status_ok.h"
+#include "stdlib/complex/float32.h"
```
Comment (Pranavchiku) [inline lib/node_modules/@stdlib/napi/argv-complex64/include/stdlib/napi/argv_complex64.h]: why do we need this?

File: lib/node_modules/@stdlib/napi/argv-complex64/include/stdlib/napi/argv_complex64.h [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+*
+* @param env      environment under which the function is invoked
+* @param name     output variable name for the array
+* @param argv     name of the variable containing add-on callback arguments
+* @param index    argument index
+*
+* @example
+* #include "stdlib/napi/argv_complex64.h"
+* #include "stdlib/napi/argv.h"
+* #include "stdlib/complex/float32.h"
```
Comment (Pranavchiku) [inline lib/node_modules/@stdlib/napi/argv-complex64/include/stdlib/napi/argv_complex64.h]: same question

File: lib/node_modules/@stdlib/napi/argv-complex64/include/stdlib/napi/argv_complex64.h [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+*     STDLIB_NAPI_ARGV_COMPLEX64( env, value, argv, 0 );
+*
+*     // ...
+*
+*     stdlib_complex64_t out = fcn( value );
+* }
+*/
+#define STDLIB_NAPI_ARGV_COMPLEX64( env, name, argv, index )           \
+	napi_value __STDLIB_NAPI_ARGV_COMPLEX64_ERR_ ## name;                   \
+	stdlib_complex64_t name;                                                                 \
```
Comment (Pranavchiku) [inline lib/node_modules/@stdlib/napi/argv-complex64/include/stdlib/napi/argv_complex64.h:57]: If we are using `stdlib_complex64_t` we will need to include a particular header, no?

File: lib/node_modules/@stdlib/napi/argv-complex64/lib/native.js [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+* @param {Complex64} v - input value
+* @returns {stdlib_complex64_t} input value
+*
+* @example
+* var Complex64 = require( '@stdlib/complex/float32' );
+*
+* var x = new Complex64( 5.0, 3.0 );
+*
+* wrapper( x );
+*/
```
Comment (Pranavchiku) [inline lib/node_modules/@stdlib/napi/argv-complex64/lib/native.js:41]: Specify return value.

File: lib/node_modules/@stdlib/napi/argv-complex64/manifest.json [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+			"libraries": [],
+			"libpath": [],
+			"dependencies": [
+		"@stdlib/assert/napi/status-ok",
+        "@stdlib/complex/float32",
+        "@stdlib/napi/argv"
+      ]
+		}
+	]
+}
```
Comment (Pranavchiku) [inline lib/node_modules/@stdlib/napi/argv-complex64/manifest.json:42]: incorrect indentation, mix of TAB and space.

File: lib/node_modules/@stdlib/napi/argv-complex64/src/addon.c [hunk 8]
```diff
@@ -0,0 +1,77 @@
...
+	// napi_status status = napi_create_object(env, &v);
+	// assert(status == napi_ok);
+
+	// status = napi_set_named_property(env, v, "re", out.re);
+	// assert(status == napi_ok);
+
+	// status = napi_set_named_property(env, v, "im", out.im);
+	// assert(status == napi_ok);
+
+	// return v;
```
Comment (Pranavchiku) [inline lib/node_modules/@stdlib/napi/argv-complex64/src/addon.c]: why are these commented ?

**Review by Pranavchiku — CHANGES_REQUESTED** [review]
Requested changes, once addressed this PR can be reviewed.

File: lib/node_modules/@stdlib/napi/argv-complex64/include/stdlib/napi/argv_complex64.h (hunk 3 above)
Comment (Userf1995) [reply lib/node_modules/@stdlib/napi/argv-complex64/include/stdlib/napi/argv_complex64.h]: @Pranavchiku I saw that some of argv package already used these that's why I used in my implemented package.
If  they are of no use, I will remove them

**Comment by Rejoan-Sardar** [comment]
@Pranavchiku I have addressed all the issues you mentioned in your review. Please take a look at the changes

**Comment by Rejoan-Sardar** [comment]
@Pranavchiku I have addressed all the issues you mentioned in your review. Please take a look at the changes

**Comment by Userf1995** [comment]
@Pranavchiku  I have addressed all the issues you mentioned in your review. Please take a look at the changes

File: lib/node_modules/@stdlib/napi/argv-complex64/include/stdlib/napi/argv_complex64.h [hunk 9]
```diff
@@ -0,0 +1,80 @@
...
+* @param env      environment under which the function is invoked
+* @param name     output variable name for the array
+* @param argv     name of the variable containing add-on callback arguments
+* @param index    argument index
+*
+* @example
+* #include "stdlib/napi/argv_complex64.h"
+* #include <node_api.h>
+*
+* static stdlib_complex64_t fcn( const stdlib_complex64_t ) {
```
Comment (kgryte) [inline lib/node_modules/@stdlib/napi/argv-complex64/include/stdlib/napi/argv_complex64.h:37]:
```suggestion
* static stdlib_complex64_t fcn( const stdlib_complex64_t v ) {
```

File: lib/node_modules/@stdlib/napi/argv-complex64/include/stdlib/napi/argv_complex64.h [hunk 10]
```diff
@@ -0,0 +1,80 @@
...
+*
+*     // Convert the first argument to a C type:
+*     STDLIB_NAPI_ARGV_COMPLEX64( env, value, argv, 0 );
+*
+*     // ...
+*
+*     stdlib_complex64_t out = fcn( value );
+* }
+*/
+#define STDLIB_NAPI_ARGV_COMPLEX64( env, name, argv, index )           \
```
Comment (kgryte) [inline lib/node_modules/@stdlib/napi/argv-complex64/include/stdlib/napi/argv_complex64.h:55]: As done elsewhere (see https://github.com/stdlib-js/stdlib/blob/develop/lib/node_modules/%40stdlib/napi/argv-double/include/stdlib/napi/argv_double.h), align `\` at 80 characters, except for those lines which are longer than 80 chars.

File: lib/node_modules/@stdlib/napi/argv-complex64/include/stdlib/napi/argv_complex64.h (hunk 5 above)
Comment (kgryte) [reply lib/node_modules/@stdlib/napi/argv-complex64/include/stdlib/napi/argv_complex64.h:57]: @Rejoan-Sardar You did not address this.

File: lib/node_modules/@stdlib/napi/argv-complex64/include/stdlib/napi/argv_complex64.h [hunk 11]
```diff
@@ -0,0 +1,80 @@
...
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*/
+
+#ifndef STDLIB_NAPI_ARGV_COMPLEX64_H
+#define STDLIB_NAPI_ARGV_COMPLEX64_H
+
+#include "stdlib/assert/napi/status_ok.h"
```
Comment (kgryte) [inline lib/node_modules/@stdlib/napi/argv-complex64/include/stdlib/napi/argv_complex64.h:22]: You are missing an include for `stdlib/napi/argv.h`.

File: lib/node_modules/@stdlib/napi/argv-complex64/include/stdlib/napi/argv_complex64.h [hunk 12]
```diff
@@ -0,0 +1,80 @@
...
+/**
+* Macro for converting an add-on callback argument to a single-precision complex floating-point number.
+*
+* @param env      environment under which the function is invoked
+* @param name     output variable name for the array
+* @param argv     name of the variable containing add-on callback arguments
+* @param index    argument index
+*
+* @example
+* #include "stdlib/napi/argv_complex64.h"
```
Comment (kgryte) [inline lib/node_modules/@stdlib/napi/argv-complex64/include/stdlib/napi/argv_complex64.h:34]: You are missing the same header in this example, as well. Please refer to https://github.com/stdlib-js/stdlib/blob/develop/lib/node_modules/%40stdlib/napi/argv-double/include/stdlib/napi/argv_double.h and investigate accordingly.

File: lib/node_modules/@stdlib/napi/argv-complex64/lib/native.js [hunk 13]
```diff
@@ -0,0 +1,49 @@
...
+
+
+// MAIN //
+
+/**
+* Wrapper function exposing the C API to JavaScript.
+*
+* @private
+* @param {Complex64} v - input value
+* @returns {stdlib_complex64_t} input value
```
Comment (kgryte) [inline lib/node_modules/@stdlib/napi/argv-complex64/lib/native.js:33]: This is incorrect. We don't embed C types in JavaScript JSDoc.

File: lib/node_modules/@stdlib/napi/argv-complex64/lib/native.js (hunk 6 above)
Comment (kgryte) [reply lib/node_modules/@stdlib/napi/argv-complex64/lib/native.js:41]: @Rejoan-Sardar You did not resolve this.

File: lib/node_modules/@stdlib/napi/argv-complex64/manifest.json [hunk 14]
```diff
@@ -0,0 +1,42 @@
...
+		{
+			"field": "libpath",
+			"resolve": true,
+			"relative": false
+		}
+	],
+	"confs": [
+		{
+			"src": [
+       			"./src/main.c"
```
Comment (kgryte) [inline lib/node_modules/@stdlib/napi/argv-complex64/manifest.json:28]: You have mixed TAB and space indentation. This file should be 2-space indentation.

File: lib/node_modules/@stdlib/napi/argv-complex64/src/addon.c [hunk 15]
```diff
@@ -0,0 +1,66 @@
...
+* @private
+* @param env    environment under which the function is invoked
+* @param info   callback data
+* @return       Node-API value
+*/
+static napi_value addon( napi_env env, napi_callback_info info ) {
+	STDLIB_NAPI_ARGV( env, info, argv, argc, 1 )
+	STDLIB_NAPI_ARGV_COMPLEX64( env, value, argv, 0 )
+	stdlib_complex64_t out = identity( value );
+	return out;
```
Comment (kgryte) [inline lib/node_modules/@stdlib/napi/argv-complex64/src/addon.c:48]: This is incorrect. `out` is **not** a Node-API value.

File: lib/node_modules/@stdlib/napi/argv-complex64/src/addon.c [hunk 16]
```diff
@@ -0,0 +1,66 @@
...
+*
+* @private
+* @param env    environment under which the function is invoked
+* @param info   callback data
+* @return       Node-API value
+*/
+static napi_value addon( napi_env env, napi_callback_info info ) {
+	STDLIB_NAPI_ARGV( env, info, argv, argc, 1 )
+	STDLIB_NAPI_ARGV_COMPLEX64( env, value, argv, 0 )
+	stdlib_complex64_t out = identity( value );
```
Comment (kgryte) [inline lib/node_modules/@stdlib/napi/argv-complex64/src/addon.c:47]: You should also mirror what we did in https://github.com/stdlib-js/stdlib/blob/develop/lib/node_modules/%40stdlib/napi/argv-double/src/addon.c#L47. Namely, this file is intended to verify that we can successfully extract a complex number object. As such, we will be passing in a known value and we should assert that it has expected values for real and imaginary components. You can use `@stdlib/complex/realf` and `@stdlib/complex/imagf` to retrieve the individual components.

File: lib/node_modules/@stdlib/napi/argv-complex64/src/main.c [hunk 17]
```diff
@@ -0,0 +1,117 @@
...
+* }
+*/
+napi_status stdlib_napi_argv_complex64( const napi_env env, const napi_value value, stdlib_complex64_t *out, const char *message, napi_value *err ) {
+
+	stdlib_assert_napi_value_is_type( env, value, napi_object, message, err );
+	if ( *err != NULL ) {
+		return napi_ok;
+	}
+
+	bool hprop;
```
Comment (kgryte) [inline lib/node_modules/@stdlib/napi/argv-complex64/src/main.c:64]: You're missing the `stdbool.h` header in the list of includes above.

File: lib/node_modules/@stdlib/napi/argv-complex64/src/main.c [hunk 18]
```diff
@@ -0,0 +1,117 @@
...
+	status = napi_has_named_property( env, value, "re", &hprop );
+	assert( status == napi_ok );
+	if ( !hprop ) {
+		status = napi_throw_type_error( env, NULL, "Invalid argument. The Complex64 object must have a real component." );
```
Comment (kgryte) [inline lib/node_modules/@stdlib/napi/argv-complex64/src/main.c:68]: Please follow error message conventions.

File: lib/node_modules/@stdlib/napi/argv-complex64/src/main.c [hunk 19]
```diff
@@ -0,0 +1,117 @@
...
+	napi_valuetype imtype;
+	status = napi_typeof( env, im0, &imtype );
+	assert( status == napi_ok );
+	if ( imtype != napi_number ) {
+		status = napi_throw_type_error( env, NULL, "Invalid argument. The Complex64 object must have an imaginary component which a number." );
+		assert( status == napi_ok );
+		return NULL;
+	}
+
+	single real;
```
Comment (kgryte) [inline lib/node_modules/@stdlib/napi/argv-complex64/src/main.c:107]: It is clear to me that you did not actually test this implementation. The C type should be `float`. Similarly, there is no `napi_get_value_single` API.

**Review by kgryte — CHANGES_REQUESTED** [review]
This PR needs considerable work before it can be merged. I left an initial round of comments.

**Comment by kgryte** [comment]
@Rejoan-Sardar Please ensure you have locally tested before you indicate that a PR is ready for review and please only mark conversations as "resolved" if you have actually resolved the comments.

**Comment by Userf1995** [comment]
Thanks for pointing this out @kgryte

File: lib/node_modules/@stdlib/napi/argv-complex64/src/main.c (hunk 19 above)
Comment (Userf1995) [reply lib/node_modules/@stdlib/napi/argv-complex64/src/main.c:107]: @kgryte This was my first PR and on that time I'm not aware lots of thing. After that I'm not go through it now I will revisit it and change accordingly.

**Pull request closed without merging**